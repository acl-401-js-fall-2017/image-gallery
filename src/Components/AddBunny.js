import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

export default class AddBunny extends PureComponent {
  render() {
    const { handleSubmit } = this.props;
    return (
      <form onSubmit={event => {
        event.preventDefault();
        const { elements } = event.target;
        const bunnyObj = { title: elements.title.value, description: elements.description.value, url: elements.url.value };
        handleSubmit(bunnyObj);
        elements.title.value = '';
        elements.description.value = '';
        elements.url.value = '';
      }}>
        <label style={{ padding: '20px' }}>
          Title:
          <input name="title"/>
        </label>
        <label style={{ padding: '20px' }}>
          Description:
          <input name="description"/>
        </label>
        <label style={{ padding: '20px' }}>
          URL:
          <input name="url"/>
        </label>
        <button type="submit">Add Bunny</button>
      </form>
    );
  }
}

AddBunny.propTypes = {
  handleSubmit: PropTypes.func
};