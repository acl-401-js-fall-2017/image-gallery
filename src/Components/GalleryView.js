import React, { PureComponent } from 'react';

export default class GalleryView extends PureComponent {
  constructor() {
    super();
    this.state = {
      i: 0,
    };
  }

  handleIndexChange(value, bunnies) {
    this.setState((prevState) => {
      const forwardRound = prevState.i === bunnies.length -1;
      let i = value === 'Next' ? prevState.i+1 : prevState.i-1;
      // let k = value === 'Previous' ? prevState.k-1 : prevState.k+1;
      const state = forwardRound ? { i: 0 } : { i }; 
      return state;
    });
  }

  render() {
    const { bunnies } = this.props;
    const bunny = bunnies[this.state.i];
    const currentImage =  
        (<div>
          <img src={bunny.url} alt="" width="40%"/>
          <p>{bunny.title}</p>
        </div>);

    return (
      <div>
        {currentImage}
        <input type="button" value="Previous" onClick={(event) => this.handleIndexChange(event.target.value, bunnies)}/>
        <input type="button" value="Next" onClick={(event) => this.handleIndexChange(event.target.value, bunnies)}/>
      </div>
    );
  }
}