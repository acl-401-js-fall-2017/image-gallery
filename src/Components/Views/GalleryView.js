import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

export default class GalleryView extends PureComponent {
  constructor() {
    super();
    this.state = {
      i: 0,
    };
  }

  handleIndexChange(value, bunnies) {
    this.setState((prevState) => {
      const isFinalIndex = prevState.i === bunnies.length -1;
      const isFirstIndex = prevState.i === 0;
      let i = value === 'Next' ? 
        (
          isFinalIndex ? 0 : prevState.i+1
        ) : (
          
          isFirstIndex ? bunnies.length-1 : prevState.i-1
        );
      const state = { i }; 
      return state;
    });
  }

  render() {
    const { bunnies, handleDelete } = this.props;
    const bunny = bunnies[this.state.i];
    const currentImage =  
        (<div>
          <img src={bunny.url} alt="" width="40%"/>
          <p>{bunny.title}</p>
        </div>);

    return (
      <div>
        {currentImage}
        <input type="button" value="Previous" onClick={(event) => this.handleIndexChange(event.target.value, bunnies)}/>
        <input type="button" value="Next" onClick={(event) => this.handleIndexChange(event.target.value, bunnies)}/>
        <button className="remove" data-value={bunny._id} onClick={({ target }) => handleDelete(target.dataset.value)}>remove</button>
      </div>
    );
  }
}

GalleryView.propTypes = {
  bunnies: PropTypes.array,
  i: PropTypes.number
};