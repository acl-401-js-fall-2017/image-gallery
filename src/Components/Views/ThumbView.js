import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

export default class ThumbView extends PureComponent {
  render() {
    const { bunnies } = this.props;

    return (
      <div>
        {bunnies.map((bunny, i) => {
          return (
            <div>
              <li key={i}>
                <figcaption><a href={bunny.url}>{bunny.title}:</a></figcaption>
                <img src={bunny.url} alt="" height="25%" width="25%"/>
              </li>
            </div>
          );
        })}
      </div>
    );
  }
}

ThumbView.propTypes = {
  bunnies: PropTypes.array
};