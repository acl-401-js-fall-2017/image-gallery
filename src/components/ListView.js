import React, { Component } from 'react';

export default class List extends Component {

  render() {
    return(
      <section>
        <ul>
          < li > Panda 1 < a href="http://f.cl.ly/items/3g3J1G0w122M360w380O/3726490195_f7cc75d377_o.jpg" >
            http: //f.cl.ly/items/3g3J1G0w122M360w380O/3726490195_f7cc75d377_o.jpg  </a></li>
          < li > Panda 2 < a href="http://static.boredpanda.com/blog/wp-content/uploads/2015/09/cute-bunnies-25__605.jpg" >
            http: //static.boredpanda.com/blog/wp-content/uploads/2015/09/cute-bunnies-25__605.jpg  </a></li>
          < li > Panda 3 < a href="http://static.boredpanda.com/blog/wp-content/uploads/2015/09/cute-bunnies-110__605.jpg" >
            http: //static.boredpanda.com/blog/wp-content/uploads/2015/09/cute-bunnies-110__605.jpg  </a></li>
        </ul>   
      </section>
    );
  }
}
