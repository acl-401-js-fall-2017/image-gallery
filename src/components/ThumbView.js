import React, { Component } from 'react';

export default class Thumb extends Component {

  render() {
    const { images, onDelete } = this.props;
    return(
      <section>
        <ul>
          {images.map((image) => {
            return (
              <li key={image._id}> {image.title} <button onClick={() => onDelete(image._id)}>X</button>
                <img className="thumb" src={image.url} alt="panda" />
              </li>);
          })}
        </ul>   
      </section>
    );
  }
}
