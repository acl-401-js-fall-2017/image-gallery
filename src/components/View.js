import React, { Component } from 'react';
import { Route, Switch, Redirect, Link } from 'react-router-dom';

import Bunnies from '../images/bunnies';
import List from './List';
import Thumbnail from './Thumbnail';
import Gallery from './Gallery';

import { addImage, removeImage } from '../data/actions';


export default class View extends Component {

  constructor() {
    super();
    this.state = {
      bunnies: Bunnies,
      // viewSelection: 'list'
    };
  }

  handleViewChange(value) {
    this.setState({
      viewSelection: value
    });
  }

  handleRemove = id => {
    const newState = removeImage(this.state, id);
    this.setState(newState);
  }

  handleAdd = image => {
    const newState = addImage(this.state, image);
    this.setState(newState);
  }


  render() {

    const displayView = {

      // home:<Home/>,

      list: <List bunnies={this.state.bunnies}
        removeImage={imageId => this.handleRemove(imageId)}
        addImage={image => this.handleAdd(image)}
      />,

      thumbnail: <Thumbnail bunnies={this.state.bunnies}/>,

      gallery: <Gallery bunnies={this.state.bunnies}/>

    };

    // const { viewSelection } = this.state;
    // let displayView;

    // (viewSelection === 'list') && (displayView = <List bunnies={this.state.bunnies} 
    //   removeImage={imageId => this.handleRemove(imageId)}
    //   addImage={image => this.handleAdd(image)}
    // />);
    // (viewSelection === 'thumbnail') && (displayView = <Thumbnail bunnies={this.state.bunnies}/>);
    // (viewSelection === 'gallery') && (displayView = <Gallery bunnies={this.state.bunnies}/>);
      

    return (
      <div>
        <div>
          <Switch>
            <Route exact path='/' component={Home}/>
            <Route exact path='/list' render={() => displayView.list}/>
            {/* <Route exact path='/list' component={List}/> */}
            <Route exact path='/thumbnail' render={() => displayView.thumbnail}/>
            {/* <Route exact path='/thumbnail' component={Thumbnail}/> */}
            <Route exact path='/gallery' render={() => displayView.gallery}/>
            {/* <Route exact path='/gallery' component={Gallery}/> */}
            <Redirect to="/"/>
          </Switch>
        </div>
      </div>
    //   <div>
    //     <input type="button" value="list" onClick={({ target }) => this.handleViewChange(target.value)}/>
    //     <input type="button" value="thumbnail" onClick={({ target }) => this.handleViewChange(target.value)}/>
    //     <input type="button" value="gallery" onClick={({ target }) => this.handleViewChange(target.value)}/>
    //     { displayView }
    //   </div> 
    );
  }
}

const Home = () => (
  <div>
    <h1>Welcome to the Image Gallery Home page!</h1>
    <Link to="/images">View Images</Link>
  </div>
);