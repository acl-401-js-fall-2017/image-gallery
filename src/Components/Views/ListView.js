import React, { PureComponent } from 'react';
import AddBunny from '../Views/AddBunny';
import PropTypes from 'prop-types';

export default class ListView extends PureComponent {
  render() {
    const { bunnies, handleSubmit, handleDelete } = this.props;
    const images = bunnies.map((bunny, i) => {
      return (
        <tr key={i}>
          <td> <a href={bunny.url}>{bunny.title}:</a> {bunny.description}</td>
          <td> 
            <button className="remove" data-value={bunny._id} onClick={({ target }) => handleDelete(target.dataset.value)}>remove</button>
          </td>
        </tr>
      );
    });
    
    return (
      <div>
        <table>
          <tbody className="images">
            {images} 
          </tbody>
        </table>
        <AddBunny handleSubmit={handleSubmit}/>
        <form onSubmit={handleSubmit}></form>
      </div>
    );
  }
}

ListView.propTypes = {
  handleSubmit: PropTypes.func,
  handleDelete: PropTypes.func,
  bunnies: PropTypes.array
};