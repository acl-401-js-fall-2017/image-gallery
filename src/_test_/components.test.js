import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import AddBunny from '../Components/AddBunny';
import View from '../View';
import ListView from '../Components/ListView';
import GalleryView from '../Components/GalleryView';
import ThumbView from '../Components/ThumbView';

const bunnies = [
  { 
    _id: '1',
    title: 'Halloween Bunny',
    description: 'Trick or treat?',
    url: 'http://f.cl.ly/items/3g3J1G0w122M360w380O/3726490195_f7cc75d377_o.jpg' },
  { 
    _id: '2',
    title: 'Belly Bunny',
    description: 'I\'m a rub-my-belly bunny',
    url: 'http://static.boredpanda.com/blog/wp-content/uploads/2015/09/cute-bunnies-25__605.jpg' },
  { 
    _id: '3',
    title: 'Side-Eye Bunny',
    description: 'Were you taking to me?',
    url: 'http://static.boredpanda.com/blog/wp-content/uploads/2015/09/cute-bunnies-110__605.jpg' }
];

it('renders adding a new bunny snapshot', () => {
  const wrapper = shallow(<AddBunny/>);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

it('renders a View component snapshot', () => {
  const wrapper = shallow(<View/>);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

it('renders a list view snapshhot', () => {
  const wrapper = shallow(<ListView bunnies={bunnies}/>);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

it('renders a gallery view snapshot', () => {
  const wrapper = shallow(<GalleryView bunnies={bunnies}/>);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

it('renders a thumbnail view snapshot', () => {
  const wrapper = shallow(<ThumbView bunnies={bunnies}/>);
  expect(toJSON(wrapper)).toMatchSnapshot();
});
