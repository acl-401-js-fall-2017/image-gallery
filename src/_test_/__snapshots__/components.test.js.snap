// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a View component snapshot 1`] = `
<div>
  <div>
    <Switch>
      <Route
        path="/list"
        render={[Function]}
      />
      <Route
        path="/gallery"
        render={[Function]}
      />
      <Route
        path="/thumbnail"
        render={[Function]}
      />
    </Switch>
  </div>
</div>
`;

exports[`renders a gallery view snapshot 1`] = `
<div>
  <div>
    <img
      alt=""
      src="http://f.cl.ly/items/3g3J1G0w122M360w380O/3726490195_f7cc75d377_o.jpg"
      width="40%"
    />
    <p>
      Halloween Bunny
    </p>
  </div>
  <input
    onClick={[Function]}
    type="button"
    value="Previous"
  />
  <input
    onClick={[Function]}
    type="button"
    value="Next"
  />
</div>
`;

exports[`renders a list view snapshhot 1`] = `
<div>
  <table>
    <tbody
      className="images"
    >
      <tr
        key="0"
      >
        <td>
           
          <a
            href="http://f.cl.ly/items/3g3J1G0w122M360w380O/3726490195_f7cc75d377_o.jpg"
          >
            Halloween Bunny
            :
          </a>
           
          Trick or treat?
        </td>
        <td>
          <button
            className="remove"
            data-value="1"
            onClick={[Function]}
          >
            remove
          </button>
        </td>
      </tr>
      <tr
        key="1"
      >
        <td>
           
          <a
            href="http://static.boredpanda.com/blog/wp-content/uploads/2015/09/cute-bunnies-25__605.jpg"
          >
            Belly Bunny
            :
          </a>
           
          I'm a rub-my-belly bunny
        </td>
        <td>
          <button
            className="remove"
            data-value="2"
            onClick={[Function]}
          >
            remove
          </button>
        </td>
      </tr>
      <tr
        key="2"
      >
        <td>
           
          <a
            href="http://static.boredpanda.com/blog/wp-content/uploads/2015/09/cute-bunnies-110__605.jpg"
          >
            Side-Eye Bunny
            :
          </a>
           
          Were you taking to me?
        </td>
        <td>
          <button
            className="remove"
            data-value="3"
            onClick={[Function]}
          >
            remove
          </button>
        </td>
      </tr>
    </tbody>
  </table>
  <AddBunny />
  <form />
</div>
`;

exports[`renders a thumbnail view snapshot 1`] = `
<div>
  <div>
    <li
      key="0"
    >
      <figcaption>
        <a
          href="http://f.cl.ly/items/3g3J1G0w122M360w380O/3726490195_f7cc75d377_o.jpg"
        >
          Halloween Bunny
          :
        </a>
      </figcaption>
      <img
        alt=""
        height="25%"
        src="http://f.cl.ly/items/3g3J1G0w122M360w380O/3726490195_f7cc75d377_o.jpg"
        width="25%"
      />
    </li>
  </div>
  <div>
    <li
      key="1"
    >
      <figcaption>
        <a
          href="http://static.boredpanda.com/blog/wp-content/uploads/2015/09/cute-bunnies-25__605.jpg"
        >
          Belly Bunny
          :
        </a>
      </figcaption>
      <img
        alt=""
        height="25%"
        src="http://static.boredpanda.com/blog/wp-content/uploads/2015/09/cute-bunnies-25__605.jpg"
        width="25%"
      />
    </li>
  </div>
  <div>
    <li
      key="2"
    >
      <figcaption>
        <a
          href="http://static.boredpanda.com/blog/wp-content/uploads/2015/09/cute-bunnies-110__605.jpg"
        >
          Side-Eye Bunny
          :
        </a>
      </figcaption>
      <img
        alt=""
        height="25%"
        src="http://static.boredpanda.com/blog/wp-content/uploads/2015/09/cute-bunnies-110__605.jpg"
        width="25%"
      />
    </li>
  </div>
</div>
`;

exports[`renders adding a new bunny snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <label
    style={
      Object {
        "padding": "20px",
      }
    }
  >
    Title:
    <input
      name="title"
    />
  </label>
  <label
    style={
      Object {
        "padding": "20px",
      }
    }
  >
    Description:
    <input
      name="description"
    />
  </label>
  <label
    style={
      Object {
        "padding": "20px",
      }
    }
  >
    URL:
    <input
      name="url"
    />
  </label>
  <button
    type="submit"
  >
    Add Bunny
  </button>
</form>
`;
